Source/DafnyLanguageServer.Test/Lookup/ReferencesTest.cs

```diff
@@ -0,0 +1,278 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Dafny.LanguageServer.IntegrationTest.Extensions;
+using Microsoft.Dafny.LanguageServer.IntegrationTest.Util;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Progress;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup {
+  public class ReferencesTest : ClientBasedLanguageServerTest {
+    private IRequestProgressObservable<IEnumerable<Location>, LocationContainer> RequestReferences(
+      TextDocumentItem documentItem, Position position) {
+      return client.RequestReferences(
+        new ReferenceParams {
+          TextDocument = documentItem.Uri,
+          Position = position
+        }, CancellationToken);
+    }
+
+    /// <summary>
+    /// Assert that when finding-references at each cursor position,
+    /// the client returns all ranges marked with regular spans.
+    /// </summary>
+    /// <param name="source"></param>
+    private async Task AssertReferences(string source) {
+      MarkupTestFile.GetPositionsAndRanges(
+        source, out var cleanSource, out var positions, out var expectedRangesArray);
+      var expectedRanges = new HashSet<Range>(expectedRangesArray);
+
+      var documentItem = CreateTestDocument(cleanSource);
+      await client.OpenDocumentAndWaitAsync(documentItem, CancellationToken);
+
+      foreach (var position in positions) {
+        var result = await RequestReferences(documentItem, position).AsTask();
+        var resultRanges = result.Select(location => location.Range).ToHashSet();
+        Assert.Equal(expectedRanges, resultRanges);
+      }
+    }
+
+    [Fact]
+    public async Task Const() {
+      var source = @"
+const ><c := 1
+method M1() {
+  print [>><c<];
+}
+method M2() {
+  print [>c><<];
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Var() {
+      var source = @"
+method M() {
+  var ><v := 1;
+  print [>><v<];
+  print [>><v<];
+  var v2 := [>><v<] + [>><v<];
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Method() {
+      var source = @"
+method ><M1() {
+}
+method M2() {
+  [>><M1<]();
+}
+method M3() {
+  [>><M1<]();
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Parameter() {
+      var source = @"
+method M(><x: int) {
+  print [>><x<];
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Module() {
```

Comment:
I wonder whether find references on definitions that may be visible outside of the file, should be enabled when not in project mode, since it might cause too few results to be returned leading to confusion. 

One option is to do the least amount of work, which is to have it be enabled as you have now, and then gather feedback from users.

The other option is to for each definition say whether it's local to the file or not, and only if it is to return references results outside of project mode.

Reply:
I think whether or not we return references of definitions outside of the current file, it will be true that (a) the list of references will be incomplete in the general case (and so users may be confused), and (b) we will want to somehow encourage users to adopt the project mode.

To address (b) I'll look into ways we can warn the users that results may be incomplete, and guide the users towards documentation on how to add an appropriate project file. But then under those circumstances, I'd rather give a "best effort" list of references, i.e. including some-but-maybe-not-all references/declarations outside the current file, rather than intentionally withholding them.

Reply:
For (b), I'm thinking of a notification that pops up whenever a file is opened that's not using project mode, that does include other files, and then ideally you could say "OK" or "OK and don't warn me again"